### PR TITLE
[MINI-3351] - Feat/mini 3351 back forward nav

### PR DESCRIPTION
**Pull request by rleojoseph**
# Description
SDK should expose a new delegate to the Host app to know whenever the MiniApp **canGoBack** & **canGoForward**, which will help the Hostapp to enable/disable the button.

We can use the existing MiniAppNavigationDelegate (or maybe a new delegate) for this approach.
But as of now, when we navigate to different pages in the JS SDK Sample app, WKWebView doesn't detect anything, so I am trying to use the Observer here. Let me know your feedback

## Links
[MINI-3351](https://jira.rakuten-it.com/jira/browse/MINI-3351)

# Checklist
- [x] I have read the [contributing guidelines](CONTRIBUTING.md).
- [x] I have completed the [SDK Development Learning Path](CONTRIBUTING.md#SDK-Development-Learning-Path) (if submitting a feature PR)
- [ ] I wrote/updated tests for new/changed code
- [x] I removed all sensitive data before every commit, including API endpoints and keys
- [ ] I ran `fastlane ci` without errors

### CODE REVIEW

**Review by Climbatize — COMMENTED** [review]
I think you did a great job, I have one concern though

File: MiniApp/Classes/core/Display/RealMiniAppView.swift [hunk 1]
```diff
@@ -122,6 +123,19 @@ internal class RealMiniAppView: UIView {
             webViewBottomConstraintStandalone?.isActive = false
         }
         MiniAppAnalytics.sendAnalytics(event: .open, miniAppId: miniAppId, miniAppVersion: miniAppVersion, projectId: projectId, analyticsConfig: analyticsConfig)
+        observeWebView()
+    }
+
+    func observeWebView() {
+        urlObservation = webView.observe(\.url, changeHandler: { (webView, _) in
+            if webView.url?.absoluteString == self.getMiniAppBaseURL(miniAppId: self.miniAppId ?? "") {
+                self.navigationDelegate?.canMiniAppNavigateTo(back: false, forward: true)
```
[the remark this answers is not in the document]
Comment (Climbatize) [reply MiniApp/Classes/core/Display/RealMiniAppView.swift:132]: can we try something like that?
```swift
 webView.observe(\.canGoBack, changeHandler: { (webView, change) in
            // check changes and call delegate
        })
```

[the remark this answers is not in the document]
Comment (rleojoseph) [reply MiniApp/Classes/core/Display/RealMiniAppView.swift:132]: I did try this approach. But the block is not called after the MiniApp is launched. Not sure why. Maybe it has a weak reference.

[the remark this answers is not in the document]
Comment (laubengaier) [reply MiniApp/Classes/core/Display/RealMiniAppView.swift:132]: @leorjoseph did you check the option parameter `NSKeyValueObservingOptions` there are a couple options like `NSKeyValueObservingOptionInitial` not sure tho but could be helpful?

**Review by laubengaier — APPROVED** [review]
lgtm so far 👍  
just a comment regarding the observation

File: MiniApp/Classes/core/Display/RealMiniAppView.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (Climbatize) [reply MiniApp/Classes/core/Display/RealMiniAppView.swift:132]: @rleojoseph by the way you should keep a reference to the result like `kvoref = webView.observe...` and in the class `deinit` method call `kvoref.invalidate()`

As for the`\.canGoBack` not triggering anything it's really inconvenient indeed. I made a test by keeping the reference as I just suggested, and everything seemed ok.

As for the improvement @laubengaier suggests, the usual observation is to observe if the variable gets a new value with the option `.new` :

```swift
webviewObserver = webView.observe(\.canGoBack, options: .new) { (label, change) in
            print("test")
        }
```

[the remark this answers is not in the document]
Comment (rleojoseph) [reply MiniApp/Classes/core/Display/RealMiniAppView.swift:132]: Thanks @Climbatize ,

I was trying to track the URL but this approach looks better & clean. Updated changes now. 

But in this approach, we need to observe `.canGoForward` as well, when the user taps on the forward button in the navigation bar. `.canGoBack` block will not be notified and we can't check if the Webview `canGoForward`, so we have to use both `.canGoBack` & `.canGoForward`

@laubengaier thanks. Maybe the `NSKeyValueObservingOptions` could have been useful in my previous approach. But with this new approach, I see we don't need to use the options.

File: MiniApp/Classes/core/Display/MiniAppNavigationConfig.swift [hunk 2]
```diff
@@ -73,6 +79,8 @@ public extension MiniAppNavigationDelegate {
     }
     func miniAppNavigation(delegate: MiniAppNavigationBarDelegate) {
     }
+    func canMiniAppNavigateTo(back: Bool, forward: Bool) {
```
Comment (Climbatize) [inline MiniApp/Classes/core/Display/MiniAppNavigationConfig.swift:82]: can you rename this to `miniAppNavigationCanGo(back: Bool, forward: Bool)`

**Review by Climbatize — APPROVED** [review]
Good job, a nice addition to the SDK. Just a little refactoring and it will be perfect

File: MiniApp/Classes/core/Display/MiniAppNavigationConfig.swift (hunk 2 above)
Comment (Climbatize) [inline MiniApp/Classes/core/Display/MiniAppNavigationConfig.swift:82]:
```suggestion
    func miniAppNavigationCanGo(back: Bool, forward: Bool) {
```

File: MiniApp/Classes/core/Display/RealMiniAppView.swift [hunk 3]
```diff
@@ -122,6 +124,16 @@ internal class RealMiniAppView: UIView {
...
+        observeWebView()
+    }
+
+    func observeWebView() {
+        canGoBackObservation = webView.observe(\.canGoBack, options: .initial) { (webView, _) in
+            self.navigationDelegate?.canMiniAppNavigateTo(back: webView.canGoBack, forward: webView.canGoForward)
+        }
+        canGoForwardObservation = webView.observe(\.canGoForward) { (webView, _) in
+            self.navigationDelegate?.canMiniAppNavigateTo(back: webView.canGoBack, forward: webView.canGoForward)
+        }
```
Comment (Climbatize) [inline MiniApp/Classes/core/Display/RealMiniAppView.swift:136]:
```suggestion
        canGoBackObservation = webView.observe(\.canGoBack, options: .initial) { (webView, _) in
            self.navigationDelegate?.miniAppNavigationCanGo(back: webView.canGoBack, forward: webView.canGoForward)
        }
        canGoForwardObservation = webView.observe(\.canGoForward) { (webView, _) in
            self.navigationDelegate?.miniAppNavigationCanGo(back: webView.canGoBack, forward: webView.canGoForward)
        }
```

**Pull request merged**